[ParseableInterface] Standardize printing for accessors

harlanhaskins opened the pull request:
We want to make sure we are consistent when printing accessors.
@jrose-apple and I landed on the following table for disambiguating all
the possible combinations of accessors:

### Computed Properties

| Semantics  | Syntax                 |
|------------|------------------------|
| Read-only  | `var x: T { get }`       |
| Read-write | `var x: T { get set }`   |

### Stored Properties

| Semantics                | Syntax                              |
|--------------------------|-------------------------------------|
| Immutable                | `let x: T`                            |
| Mutable (trivial setter) | `var x: T`                            |
| Mutable (custom setter)  | `@_hasStorage var x: T { get set }`   |
| Private(set)             | `@_hasStorage var x: T { get }`       |

This patch introduces the `@_hasStorage` attribute which is printed for declarations that have storage and need to print custom accessors.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -838,6 +843,11 @@ static bool hasNonMutatingSetter(const AbstractStorageDecl *ASD) {
 
+static bool isSimpleStoredWithPublicSetter(const AbstractStorageDecl *asd) {
+  return asd->getImplInfo().isSimpleStored() &&
+    !asd->getAttrs().hasAttribute<SetterAccessAttr>();
```
This doesn't seem right; it could be an `open` var with a `public(set)` (though I don't exactly know if we support that), or a `public` var with a `@usableFromInline` setter (though I don't know if we support *that),* or even a `public` var with a `public(set)` (which is only a warning).

harlanhaskins replied:
> it could be an `open` var with a `public(set)`

We do currently support this, so I'm gonna check for non-public-or-usableFromInline in the AccessLevel

> a `public` var with a `@usableFromInline` setter

Technically we support this on internal computed properties, but those won't get printed in interfaces at all.

> or even a `public` var with a `public(set)`

😢

harlanhaskins replied:
The check I'm doing now is if the setter's access level is less than the access level of the storage decl itself. Is that right?

jrose-apple commented on test/ParseableInterface/access-filter.swift:
```diff
@@ -109,7 +109,7 @@ extension UFIProto {
 // CHECK: extension PublicStruct {{[{]$}}
 extension PublicStruct {
-  // CHECK: public private(set) static var secretlySettable: Int{{$}}
+  // CHECK: @_hasInitialValue public static var secretlySettable: Int
```
We want to check that it ends in `{ get }` now, right?

harlanhaskins replied:
Yep, I'm working on this test now

jrose-apple reviewed (commented):
Can you add some IRGen tests to demonstrate that we're computing layout correctly in all cases?

harlanhaskins commented:
@jrose-apple For the IRGen tests, should I do it as a SIL test instead since this attribute doesn't exist outside of interfaces and interfaces don't reach IRGen?

harlanhaskins commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -4699,6 +4699,42 @@ static void diagnoseAndIgnoreObservers(Parser &P,
   }
 }
 
+static StorageImplInfo classifyWithHasStorageAttr(
```
@rjmccall This patch introduces a `@_hasStorage` attribute that allows us to break an ambiguity in parseable interfaces.

The mapping we've settled on is an understanding that
```swift
public private(set) var x: Int
```
maps to
```swift
@_hasStorage public var x: Int { get }
```
and
```swift
public var x: Int {
  willSet/didSet {}
}
```
maps to
```swift
@_hasStorage public var x: Int { get set }
```

With that in mind, is this mapping correct?

rjmccall replied:
You should be able to rename the `@_sil_stored` attribute for this purpose.

If you're changing this, could you change it to list/expect all the public accessors, i.e. including `modify` and (eventually) `_read`?  You might need to (1) move the logic for deciding the opaque accessors into Sema and then (2) change `StorageImplInfo` to store that bitset.

harlanhaskins replied:
I'm not sure how to handle `_modify` and `_read` if `@_hasStorage` is provided.

What are the semantics of `@sil_stored var x: T { get set }` vs. just `@sil_stored var x: T`? Same for `@sil_stored let x: T { get }` vs. just `@sil_stored let x: T`?

rjmccall replied:
Currently it declares that the accessors are actually available.  I'm not sure that's important to know in SIL.  I do think there's a strong argument for it being useful to unambiguously list out the available opaque accessors, but it should be the full set, not just `{get set}` and then inferring `modify` and `read` and so on.  However, to express that properly, you might have to change the representation a bit.

harlanhaskins replied:
I'm going to do this in a future patch, because testing parseable interfaces is currently blocked on this representational change. For now, there's a loud FIXME explaining how it should eventually work.

rjmccall replied:
Sounds good.

jrose-apple commented:
> For the IRGen tests, should I do it as a SIL test instead since this attribute doesn't exist outside of interfaces and interfaces don't reach IRGen?

I was thinking that you'd IRGen a client that imports an interface.

harlanhaskins commented:
@jrose-apple Does the IRGen test look good?

harlanhaskins commented on test/SILOptimizer/accessed_storage_analysis.sil:
```diff
@@ -340,7 +340,7 @@ class C {
 
 // CHECK-LABEL: @readIdentifiedClass
 // CHECK: [read] Class %0 = argument of bb0 : $C
-// CHECK: Field: @_hasStorage var property: Int
```
These are changed because this printer doesn't print for SIL. I could tell it to print for SIL, but that would cause other tests to print slightly differently, and this output isn't intended to convey full semantics.

harlanhaskins commented:
Addresses rdar://43815861

jrose-apple commented on test/ParseableInterface/stored-properties-client.swift:
```diff
@@ -0,0 +1,144 @@
+// RUN: %target-swift-frontend -emit-ir %s -I %t -enable-parseable-module-interface | %FileCheck %s -check-prefix CHECK -check-prefix COMMON
+
+// RUN: %target-swift-frontend -typecheck %S/stored-properties.swift -enable-resilience -module-name StoredProperties -emit-parseable-module-interface-path %t/StoredProperties.swiftinterface
+// RUN: %target-swift-frontend -emit-ir %s -I %t -enable-parseable-module-interface | %FileCheck %s -check-prefix RESILIENT -implicit-check-not CHECK -check-prefix COMMON
```
`-implicit-check-not` adds a negative pattern, not a negative prefix. What do you actually want to check here?

jrose-apple replied:
I don't think there's a way to do that today, but I think it's okay for this test.

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on test/ParseableInterface/stored-properties-client.swift:
```diff
@@ -12,7 +12,7 @@ import StoredProperties
 /// .swiftinterface file, in both resilient and non-resilient cases.
 
-// COMMON: %[[BAGOFVARIABLES:T16StoredProperties14BagOfVariablesV]] = type <{ %TSi, %TSb, [7 x i8], %TSi }>
+// COMMON: %[[BAGOFVARIABLES:T16StoredProperties14BagOfVariablesV]] = type <{ %TSi, %TSb, [{{[0-9]}} x i8], %TSi }>
```
This is the thing we're supposed to be checking, though…can you do it as {{7|9}} or something like that? Or use FileCheck's fancy new -D feature?

harlanhaskins replied:
I think we'd want it to be {{3|7}}, because it's going to pad to word boundaries after a Bool.

jrose-apple replied:
I just made up numbers, go for it.

harlanhaskins committed:
[ParseableInterface] Standardize printing for accessors

We want to make sure we are consistent when printing accessors.
@jrose-apple and I landed on the following table for disambiguating all
the possible combinations of accessors:

\### Computed Properties

| Semantics  | Syntax                 |
|------------|------------------------|
| Read-only  | var x: T { get }       |
| Read-write | var x: T { get set }   |

\### Stored Properties

| Semantics                | Syntax                              |
|--------------------------|-------------------------------------|
| Immutable                | let x: T                            |
| Mutable (trivial setter) | var x: T                            |
| Mutable (custom setter)  | @_hasStorage var x: T { get set }   |
| Private(set)             | @_hasStorage var x: T { get }       |

This first commit enables printing @_hasStorage outside of SIL mode.

harlanhaskins committed:
[Sema] Add implicit @_hasStorage attribute for printing

harlanhaskins committed:
[Parse] Parse @_hasStorage attribute outside of SIL

harlanhaskins committed:
[Sema] Allow explicit @_hasStorage attribute outside of SIL

harlanhaskins committed:
[ASTPrinter] Print property observers/private(set) with @_hasStorage

harlanhaskins committed:
[Serialization] Add @_hasStorage and private(set) while deserializing

harlanhaskins committed:
[ParseableInterface] Test accessor printing with @_hasStorage

harlanhaskins committed:
[test] Ensure layout and accessors are correct with @_hasStorage

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3019,21 +3019,29 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
+      if (*setterAccess < *accessLevel) {
+        AddAttribute(
+          new (ctx) SetterAccessAttr(SourceLoc(), SourceLoc(),
+                                     *setterAccess, /*implicit*/true));
```
This is unfortunate. I wish I had noticed it at the time, but it's weird to be different from the regular access.

harlanhaskins replied:
I agree that it's unfortunate, but I'm not sure how else to preserve this. The rest of the compiler assumes setter access is determined by the presence of this attribute. If there's a better way than synthesizing this here, I'm happy to change it.
